File: src/VBNetFx/Controls/ADNullableValueSubControls/ADCheckBoxedValueForNullableValueControl.vb
Imports System.Windows.Forms
Imports System.Drawing
Imports System.Collections

Friend Class ADCheckBoxedValueForNullableValueControl
    Inherits ADEditableValueForNullableValueControlTemplate(Of Boolean)

    Protected WithEvents myCheckBox As CheckBox

    Sub New()
        MyBase.New()
        myCheckBox = New CheckBox
        myControl = myCheckBox
        myControl.Name = "ADCheckBoxedValueForUVTControl" & myInstanceCounter
        myCheckBox.ThreeState = True
        myCheckBox.CheckState = CheckState.Indeterminate
        myCheckBox.Text = ""
        myCheckBox.BackColor = BackColor
        PaddingToCaption = 0
        CheckAlignment = ContentAlignment.MiddleCenter
    End Sub

    Public Overrides Property Size() As System.Drawing.Size
        Get
            Return myControl.Size
        End Get
        Set(ByVal Value As System.Drawing.Size)
            myControl.Size = Value
        End Set
    End Property

    Public Overrides Function MeasureHeight() As Integer
        Return (myControl.Height) * -1
    End Function

    Public Property PaddingToCaption() As Integer
        Get
            Return myCheckBox.Padding.Left
        End Get
        Set(ByVal value As Integer)
            myCheckBox.Padding = New System.Windows.Forms.Padding(value, _
                                    myCheckBox.Padding.Top, _
                                    myCheckBox.Padding.Right, _
                                    myCheckBox.Padding.Bottom)
        End Set
    End Property

    Public Property CheckAlignment() As ContentAlignment
        Get
            Return myCheckBox.CheckAlign
        End Get
        Set(ByVal value As ContentAlignment)
            myCheckBox.CheckAlign = value
        End Set
    End Property

    Public Overrides Property Value() As Object
        Get
            If myCheckBox.CheckState = CheckState.Indeterminate Then
                Return Nothing
            End If
            Return myCheckBox.Checked
        End Get
        Set(ByVal Value As Object)
            Dim locTOB As ADDBNullable(Of Boolean) = CType(Value, ADDBNullable(Of Boolean))
            If Not locTOB.HasValue Then
                myCheckBox.CheckState = CheckState.Indeterminate
            Else
                If CBool(CType(Value, ADDBNullable(Of Boolean)).Value) Then
                    myCheckBox.CheckState = CheckState.Checked
                Else
                    myCheckBox.CheckState = CheckState.Unchecked
                End If
            End If
        End Set
    End Property

    Private Sub myComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles myCheckBox.CheckedChanged
        OnValueChanged(e)
    End Sub
End Class

File: src/NetFxCSharp/Controls/ADNullableValueSubControls/ADCheckBoxedValueForNullableValueControl.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace ActiveDev.Controls
{
    internal class ADCheckBoxedValueForNullableValueControl : ADEditableValueForNullableValueControlTemplate<bool>
    {
        private CheckBox _myCheckBox;
        // TODO(vb-convert): WithEvents member is reassigned outside InitializeComponent; re-wiring retained.
        protected CheckBox myCheckBox
        {
            get
            {
                return _myCheckBox;
            }

            set
            {
                if (_myCheckBox != null)
                {
                    _myCheckBox.CheckedChanged -= myComboBox_SelectedIndexChanged;
                }

                _myCheckBox = value;
                if (_myCheckBox != null)
                {
                    _myCheckBox.CheckedChanged += myComboBox_SelectedIndexChanged;
                }
            }
        }

        public ADCheckBoxedValueForNullableValueControl() : base()
        {
            myCheckBox = new CheckBox();
            myControl = myCheckBox;
            myControl.Name = "ADCheckBoxedValueForUVTControl" + myInstanceCounter;
            myCheckBox.ThreeState = true;
            myCheckBox.CheckState = CheckState.Indeterminate;
            myCheckBox.Text = "";
            myCheckBox.BackColor = BackColor;
            PaddingToCaption = 0;
            CheckAlignment = ContentAlignment.MiddleCenter;
        }

        public override System.Drawing.Size Size
        {
            get
            {
                return myControl.Size;
            }

            set
            {
                System.Drawing.Size Value = value;
                myControl.Size = Value;
            }
        }

        public override int MeasureHeight()
        {
            return (myControl.Height) * -1;
        }

        public int PaddingToCaption
        {
            get
            {
                return myCheckBox.Padding.Left;
            }

            set
            {
                myCheckBox.Padding = new System.Windows.Forms.Padding(value, myCheckBox.Padding.Top, myCheckBox.Padding.Right, myCheckBox.Padding.Bottom);
            }
        }

        public ContentAlignment CheckAlignment
        {
            get
            {
                return myCheckBox.CheckAlign;
            }

            set
            {
                myCheckBox.CheckAlign = value;
            }
        }

        public override object Value
        {
            get
            {
                if (myCheckBox.CheckState == CheckState.Indeterminate)
                {
                    return null;
                }

                return myCheckBox.Checked;
            }

            set
            {
                object Value = value;
                ADDBNullable<bool> locTOB = ((ADDBNullable<bool>)Value);
                if (!(locTOB.HasValue))
                {
                    myCheckBox.CheckState = CheckState.Indeterminate;
                }
                else
                {
                    if (System.Convert.ToBoolean(((ADDBNullable<bool>)Value).Value))
                    {
                        myCheckBox.CheckState = CheckState.Checked;
                    }
                    else
                    {
                        myCheckBox.CheckState = CheckState.Unchecked;
                    }
                }
            }
        }

        private void myComboBox_SelectedIndexChanged(object sender, System.EventArgs e)
        {
            OnValueChanged(e);
        }
    }
}